Payjoin v4

arminsabouri opened the pull request.

arturgontijo committed:
[PDK] Initial approach.

arturgontijo committed:
[PDK] Use rust-payjoin branch that accepts AddressType::P2wsh

arturgontijo committed:
[PDK] Stable sender workflow.

arturgontijo committed:
[PDK] payjoin mod.

arturgontijo committed:
[PDK] Use try_preserving_privacy()

arturgontijo committed:
[PDK] Use SLE

arturgontijo committed:
[PDK] SEL working...

arturgontijo committed:
Merge branch 'refs/heads/master' into payjoin-v3

# Conflicts:
#	liana-gui/src/app/message.rs
#	liana-gui/src/app/state/receive.rs
#	liana-gui/src/app/view/receive.rs
#	liana-gui/src/daemon/client/mod.rs
#	liana-gui/src/daemon/embedded.rs
#	liana-gui/src/daemon/mod.rs
#	liana-gui/src/services/connect/client/backend/mod.rs

arturgontijo committed:
[PDK] Sync master + OHttpKeys in DB.

arturgontijo committed:
[PDK] Better receiver code structure + clippy.

arturgontijo committed:
[PDK] fix

arturgontijo committed:
Code review

arturgontijo committed:
[PDK] Use SEL remote branch + contribute_inputs_with_weights()

arturgontijo committed:
Using SEL with p2wsh hack.

arminsabouri committed:
Impl payjoin methods on lianad

arminsabouri committed:
Should save amount and address from bip21

arminsabouri committed:
Remove redundant results in db interface

arminsabouri committed:
Handle ohttp and finalize sender psbt error in rpc endpoints

arminsabouri committed:
Run project wide fmt

arminsabouri committed:
Remove get payjoin info endpoint

arminsabouri committed:
Simplify status to wrap  only one enum

arminsabouri committed:
Remove unused receiver sender status enums

benalleng committed:
Cleanup receiver unwraps and error propogation

Just a small commit to begin cleaning up some unhandled unwraps and
begin to move away from the reliance on the session metadata.

arminsabouri committed:
Refactor pj db tables to use seperate table for events

arminsabouri committed:
Remove unused pj json-rpc endpoints

arminsabouri committed:
Revert "Remove get payjoin info endpoint"

This reverts commit 357138b2b2c6649fe64691d4a1c505ee47a278c9.

arminsabouri committed:
Do update existing pj proposal

arminsabouri committed:
Fix getting active payjoins sessions

arminsabouri committed:
gui: path is ready if all owned inputs are signed

arminsabouri committed:
Update pdk branch to use-witness-for-expected-weight

This also updated the receiver to derive input weight when constructing
inputs

arminsabouri committed:
Add expected weight to inputs

arminsabouri committed:
Resolve clippy linting warnings

And cleaned up a unneeded TODO comment

benalleng committed:
[WIP] add somre receiver state checks

Directly reference data made available within liana and do not actually
use the closure that calls out to rust-payjoin.

arminsabouri committed:
Save psbt with fees applied

arminsabouri committed:
Refactor schema to remove redundant session id

arminsabouri committed:
Update payjoin persistance related test utils

benalleng committed:
Add insert_inputs_seen_before database method

liana did not have a way to add coins from external wallets to the db,
this adds those coins to the db and forces is_from_self to be false and
sets the wallet_id to 2 which is never the users wallet as it is
hardcoded to 1.

arminsabouri committed:
hack: workaround getting payjoin status

arminsabouri committed:
Add unit tests for payjoin db utils

arminsabouri committed:
Rename address struct payjoin-uri -> bip21

arminsabouri committed:
Replace bip21 string with url type

benalleng committed:
Add support for bip21 qr codes on the receiver

This adds support for displaying bip21 qr codes to the receiver when
initializing a payjoin.

arminsabouri committed:
Breakout payjoin status from bip21

arminsabouri committed:
Get payjoin info if spend is original proposal or ready to be signed psbt

arminsabouri committed:
Rename get all session ids to get all active session ids

arminsabouri committed:
Return only payjoin status from json rpc endpoint

arminsabouri committed:
Seperate out bip21 from spendtx model

arminsabouri committed:
Clean up unused code in get payjoin info

arminsabouri committed:
Return errors from p2wsh hot signer

arminsabouri committed:
Receiver closure should be mutable refrences

arminsabouri committed:
Store original and proposed txids for pj sender

arminsabouri committed:
Store original and proposed txids for pj receiver

arminsabouri committed:
Resolve missing `has_bip21` flag

arminsabouri committed:
todo!() remaining test utils methods

arminsabouri committed:
Fix payjoin visibility in helpers.rs

arminsabouri committed:
Consolidate use of OHTTP_RELAY const

arminsabouri committed:
Consolidate use of PAYJOIN_DIR const

arminsabouri commented on lianad/src/payjoin/db.rs:
```diff
@@ -0,0 +1,149 @@
+    {
+        let mut db_conn = self.db.connection();
+        let events = db_conn.load_receiver_session_events(&self.session_id);
+        let deserialized_events: Result<Vec<_>, _> = events
+            .into_iter()
+            .map(|event| serde_json::from_slice(&event).map_err(PersisterError::Deserialize))
+            .collect();
+        Ok(Box::new(deserialized_events?.into_iter()))
```
Do we have to collect and then into_iter()?
can we just 
```rust
let iter = events.iter().map(|event| {
    serde_json::from_slice(event).map_err(PersisterError::Deserialize)
});
Ok(Box::new(iter))
```

arminsabouri commented on lianad/src/payjoin/db.rs:
```diff
@@ -0,0 +1,149 @@
+    }
+}
+
+impl SessionPersister for SenderPersister {
```
Same comments are as above apply here

arminsabouri commented on lianad/src/jsonrpc/rpc.rs:
```diff
@@ -173,6 +174,9 @@ impl From<commands::CommandError> for Error {
                 Error::new(ErrorCode::ServerError(BROADCAST_ERROR), e.to_string())
             }
+            commands::CommandError::FailedToPostOriginalPayjoinProposal(_) => {
+                Error::new(ErrorCode::ServerError(BROADCAST_ERROR), e.to_string())
```
Use different error const

arminsabouri commented on lianad/src/commands/mod.rs:
```diff
@@ -790,6 +906,7 @@ impl DaemonControl {
+            info!("Updating spend: {:?}", txid);
             let db_tx = db_psbt.unsigned_tx.clone();
             for i in 0..db_tx.input.len() {
                 if tx
```
debug statment can remove

arminsabouri commented on lianad/src/commands/mod.rs:
```diff
@@ -359,7 +379,103 @@ impl DaemonControl {
-        GetAddressResult::new(address, new_index)
+        GetAddressResult::new(address, new_index, None)
+    }
+
```
Missing error handling on all new json RPC endpoints

arminsabouri commented on lianad/src/bitcoin/poller/mod.rs:
```diff
@@ -23,8 +23,7 @@ pub struct Poller {
     secp: secp256k1::Secp256k1<secp256k1::VerifyOnly>,
-    // The receive and change descriptors (in this order).
-    descs: [descriptors::SinglePathLianaDesc; 2],
+    desc: descriptors::LianaDescriptor,
```
Can we revert these changes? Not sure why they needed to be made

arminsabouri commented on liana/src/spend.rs:
```diff
@@ -55,6 +55,8 @@ pub enum SpendCreationError {
     SanityCheckFailure(Psbt),
     FetchingTransaction(bitcoin::OutPoint),
     CoinSelection(InsufficientFunds),
+    // TODO: wrap more specific error
```
Is there actually a more specific error we can wrap? This feels like a item that came from the hackathon

arminsabouri commented on liana/src/signer.rs:
```diff
@@ -385,6 +385,7 @@ impl HotSigner {
                     &mut psbt.inputs[i],
                     i,
                 )?;
+                log::info!("Signed input at {}", i);
```
debug statment. Can remove

arminsabouri commented on liana-gui/src/services/connect/client/backend/mod.rs:
```diff
@@ -619,9 +620,22 @@ impl Daemon for BackendWalletClient {
         })
     }
 
+    async fn receive_payjoin(&self) -> Result<GetAddressResult, DaemonError> {
```
Should implement these

arminsabouri replied:
Although im unsure how to setup Liana so we use this daemon

arminsabouri commented on liana-gui/src/daemon/model.rs:
```diff
@@ -53,6 +54,9 @@ pub struct SpendTx {
     pub updated_at: Option<u32>,
     pub kind: TransactionKind,
+    pub payjoin_status: Option<PayjoinStatus>,
+    // // TODO: use a stronger type like bitcoin_uri
+    // pub bip21: String,
```
Can remove

arminsabouri commented on liana-gui/src/app/view/receive.rs:
```diff
@@ -348,6 +398,7 @@ pub fn verify_address_modal<'a>(
 }
 
 pub fn qr_modal<'a>(qr: &'a qr_code::Data, address: &'a String) -> Element<'a, Message> {
+    let max_width = if address.len() > 64 { 600 } else { 400 };
```
I think we are re-using this modal for the bip21. In which case we should rename the address to param to something more general

arminsabouri commented on liana-gui/src/app/view/psbt.rs:
```diff
@@ -418,6 +423,23 @@ pub fn spend_overview_view<'a>(
+                                    button::secondary(None, "Send Payjoin")
+                                        .on_press(Message::Spend(SpendTxMessage::SendPayjoin))
+                                        .width(Length::Fixed(150.0)),
+                                )
```
Triple else indicates a better way to write this

arminsabouri committed:
Clean up unused error fields in sendpayjoin modal

arminsabouri committed:
Remove unneeded debug logs

arminsabouri committed:
Fix visibility in payjoin sub mod

arminsabouri committed:
Remove unused db error variant

arminsabouri committed:
Lazily deserialize session events in load()

arminsabouri committed:
Add rustdocs to finalize_psbt()

benalleng committed:
Cleanup clippy warnings for CI

We have a few stray clippy warnings causing our linter to fail in the
CI.

arminsabouri committed:
Restore psbt input fields after finalizing psbt

arminsabouri committed:
Let rust-payjoin restore psbt inputs

arminsabouri committed:
Remove unused db deser error variant

benalleng committed:
Add todo insert_inputs_seen dummmy test utils method

This missing method in the dummy test utils was again causing some more
clippy faliures.

yashrajd commented on liana-gui/src/app/view/psbt.rs:
```diff
@@ -331,6 +332,10 @@ pub fn spend_header<'a>(
 }
 
+pub fn payjoin_send_success_view<'a>() -> Element<'a, Message> {
+    card::simple(text("Payjoin sent successfully")).into()
```
TL;DR 

sender CTA can read something like: "Payjoin transaction initiated! Be sure to check the PSBT tab in <appropriate number in hours> hours."

or if you don't wanna use a time at all

"Payjoin transaction initiated! Be sure to check the PSBT tab later to finalise transaction"


–––-
IIRC this isn't immediately actionable, and we don't know when it will be.

We should actually use the unix expiry time and use it to tell sender when to check back.  Actually, similar (not the same) CTA could be shown to the Receiver as well.

For e.g. sender CTA can read something like: "Payjoin transaction initiated! Be sure to check the PSBT tab in <appropriate number in hours> hours.

Ideally, we also show the sender a notification when they come back online and the client obtains the final psbt. Tapping on the notification takes them to the PSBT page.

arminsabouri replied:
>Payjoin transaction initiated! Be sure to check the PSBT tab later to finalise transaction

Yeah something like this would work. 

>Ideally, we also show the sender a notification when they come back online and the client obtains the final psbt. Tapping on the notification takes them to the PSBT page.

This would be ideal however the GUI framework is quite limited in capability. I'm sure something like this is possible.

arminsabouri committed:
Remove buggy sender fallback tx check

benalleng committed:
Process Receiver sessions is no longer halted by a failed session

The for loop in our receiver session was causing any error in a session
to bubble up and completely stop the loop meaning follow-up sessions
could not proceed forward.

benalleng committed:
Sender checks skip failed sessions

benalleng committed:
Prevent expired sessions from panicing

benalleng committed:
handle replay session events errors
